### PR TITLE
[Tokenizer Utils Base] Make pad function more flexible

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

Currently, tokenizers forces the dict to be padded to have an `input_ids` key. This restricts transformers tokenizers too much for models outside of NLP, such as Wav2Vec2: https://github.com/huggingface/transformers/pull/9659/files?file-filters%5B%5D=.py
As discussed offline, the cleanest approach is to add `input_ids` to the class attribute `model_input_names` and enforce a certain order. This is ensured by a test and a couple of comments that make the reader aware of it.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.

### CODE REVIEW

File: src/transformers/tokenization_utils_base.py [hunk 1]
```diff
@@ -1493,6 +1493,7 @@ class PreTrainedTokenizerBase(SpecialTokensMixin):
     pretrained_init_configuration: Dict[str, Dict[str, Any]] = {}
     max_model_input_sizes: Dict[str, Optional[int]] = {}
     model_input_names: List[str] = ["token_type_ids", "attention_mask"]
+    required_input_name: str = ["input_ids"]
```
Comment (patrickvonplaten) [inline src/transformers/tokenization_utils_base.py:1496]:
```suggestion
    required_input_name: str = "input_ids"
```

File: src/transformers/tokenization_utils_base.py [hunk 2]
```diff
@@ -3005,41 +3008,50 @@ def _pad(
...
+                    )
                 if "token_type_ids" in encoded_inputs:
                     encoded_inputs["token_type_ids"] = [self.pad_token_type_id] * difference + encoded_inputs[
                         "token_type_ids"
                     ]
                 if "special_tokens_mask" in encoded_inputs:
                     encoded_inputs["special_tokens_mask"] = [1] * difference + encoded_inputs["special_tokens_mask"]
-                encoded_inputs["input_ids"] = [self.pad_token_id] * difference + encoded_inputs["input_ids"]
+                encoded_inputs[self.required_input_name] = [self.pad_token_id] * difference + encoded_inputs[
+                    self.required_input_name
```
Comment (sgugger) [inline src/transformers/tokenization_utils_base.py:3049]: There is a lot of padding formatting (IMO) here all because `encoded_inputs[self.required_input_name]` is a bit long. Can we store that in some variable? (Of course is should only be replaced in the right sides, not the left side.)

**Review by sgugger — APPROVED** [review]
Thanks for doing this in a clean PR! LGTM apart from black making the code less readable.

**Review by sgugger — APPROVED** [review]
Perfect!

**Review by LysandreJik — APPROVED** [review]
This is great! Do you think we could also add a test for the updated `pad` method?

File: tests/test_tokenization_gpt2.py [hunk 3]
```diff
@@ -174,3 +174,7 @@ def test_padding(self, max_length=15):
                     max_length=max_length,
                     padding="max_length",
                 )
+
+    # tokenizer has to padding token
```
Comment (patrickvonplaten) [inline tests/test_tokenization_gpt2.py:178]:
```suggestion
    # tokenizer has no padding token
```

File: tests/test_tokenization_openai.py [hunk 4]
```diff
@@ -128,3 +128,7 @@ def test_padding(self, max_length=15):
                     max_length=max_length,
                     padding="max_length",
                 )
+
+    # tokenizer has to padding token
```
Comment (patrickvonplaten) [inline tests/test_tokenization_openai.py:132]:
```suggestion
    # tokenizer has no padding token
```

File: tests/test_tokenization_reformer.py [hunk 5]
```diff
@@ -107,6 +107,10 @@ def test_padding(self, max_length=15):
                     padding="max_length",
                 )
 
+    # tokenizer has to padding token
```
Comment (patrickvonplaten) [inline tests/test_tokenization_reformer.py:110]:
```suggestion
    # tokenizer has no padding token
```
